storage/innobase/xtrabackup/src/ds_decrypt.c

```diff
@@ -0,0 +1,845 @@
+	ulong checksum, checksum_exp;
+	ulonglong tmp;
+	xb_rcrypt_result_t result = XB_CRYPT_READ_CHUNK;
+
+	*bytes_processed = 0;
+	ptr = buf;
+
+	CHECK_BUF_SIZE(ptr, XB_CRYPT_CHUNK_MAGIC_SIZE, buf, len);
+	if (memcmp(ptr, XB_CRYPT_CHUNK_MAGIC3,
+		   XB_CRYPT_CHUNK_MAGIC_SIZE) == 0) {
+		version = 3;
+	} else if (memcmp(ptr, XB_CRYPT_CHUNK_MAGIC2,
+			  XB_CRYPT_CHUNK_MAGIC_SIZE) == 0) {
+		version = 2;
+	} else if (memcmp(ptr, XB_CRYPT_CHUNK_MAGIC1,
+			  XB_CRYPT_CHUNK_MAGIC_SIZE) == 0) {
+		version = 1;
+	} else {
+		msg("%s:%s: wrong chunk magic at offset 0x%llx.\n",
+		    my_progname, __FUNCTION__, thd->offset);
+		result = XB_CRYPT_READ_ERROR;
+		goto exit;
+	}
+
+	ptr += XB_CRYPT_CHUNK_MAGIC_SIZE;
+	thd->offset += XB_CRYPT_CHUNK_MAGIC_SIZE;
+
+	CHECK_BUF_SIZE(ptr, 8, buf, len);
+	tmp = uint8korr(ptr);	/* reserved */
+	ptr += 8;
+	thd->offset += 8;
+
+	CHECK_BUF_SIZE(ptr, 8, buf, len);
+	tmp = uint8korr(ptr);	/* original size */
+	ptr += 8;
+	if (tmp > INT_MAX) {
+		msg("%s:%s: invalid original size at offset 0x%llx.\n",
+		    my_progname, __FUNCTION__, thd->offset);
+		result = XB_CRYPT_READ_ERROR;
+		goto exit;
+	}
+	thd->offset += 8;
+	thd->to_len = (size_t)tmp;
+
+	if (thd->to_size < thd->to_len) {
+		thd->to = (uchar *) my_realloc(
+				thd->to,
+				thd->to_len,
+				MYF(MY_FAE | MY_ALLOW_ZERO_PTR));
+		thd->to_size = thd->to_len;
+	}
+
+	CHECK_BUF_SIZE(ptr, 8, buf, len);
+	tmp = uint8korr(ptr);	/* encrypted size */
+	ptr += 8;
+	if (tmp > INT_MAX) {
+		msg("%s:%s: invalid encrypted size at offset 0x%llx.\n",
+		    my_progname, __FUNCTION__, thd->offset);
+		result = XB_CRYPT_READ_ERROR;
+		goto exit;
+	}
+	thd->offset += 8;
+	thd->from_len = (size_t)tmp;
+
+	CHECK_BUF_SIZE(ptr, 4, buf, len);
+	checksum_exp = uint4korr(ptr);	/* checksum */
+	ptr += 4;
+	thd->offset += 4;
+
+	/* iv size */
+	if (version == 1) {
+		thd->iv_len = 0;
+		thd->iv = NULL;
+	} else {
+		CHECK_BUF_SIZE(ptr, 8, buf, len);
+
+		tmp = uint8korr(ptr);
+		if (tmp > INT_MAX) {
+			msg("%s:%s: invalid iv size at offset 0x%llx.\n",
+			    my_progname, __FUNCTION__, thd->offset);
+			result = XB_CRYPT_READ_ERROR;
+			goto exit;
+		}
+		ptr += 8;
+		thd->offset += 8;
+		thd->iv_len = (size_t)tmp;
+	}
+
+	if (thd->iv_len > 0) {
+		CHECK_BUF_SIZE(ptr, thd->iv_len, buf, len);
+		thd->iv = ptr;
+		ptr += thd->iv_len;
+	}
+
+	/* for version euqals 2 we need to read in the iv data but do not init
+	CTR with it */
+	if (version == 2) {
+		thd->iv_len = 0;
+		thd->iv = 0;
+	}
+
+	if (thd->from_len > 0) {
+		CHECK_BUF_SIZE(ptr, thd->from_len, buf, len);
+		thd->from = ptr;
+		ptr += thd->from_len;
+	}
+
+	xb_ad(thd->from_len <= thd->to_len);
+
+	checksum = crc32(0, thd->from, thd->from_len);
+	if (checksum != checksum_exp) {
+		msg("%s:%s invalid checksum at offset 0x%llx, "
+		    "expected 0x%lx, actual 0x%lx.\n", my_progname,
+		    __FUNCTION__, thd->offset, checksum_exp, checksum);
+		result = XB_CRYPT_READ_ERROR;
+		goto exit;
+	}
+
+	thd->offset += thd->from_len;
```

Comment:
Offset is used only in `parse_xbcrypt_chunk` in order to print error message. Incrementing it once simply doesn't work for that purpose.